include/ignition/gazebo/detail/EntityComponentManager.hh

```diff
@@ -81,11 +83,24 @@ auto CompareData = [](const DataType &_a, const DataType &_b) -> bool
 
 //////////////////////////////////////////////////
 template<typename ComponentTypeT>
-ComponentKey EntityComponentManager::CreateComponent(const Entity _entity,
+ComponentTypeT *EntityComponentManager::CreateComponent(const Entity _entity,
             const ComponentTypeT &_data)
 {
-  return this->CreateComponentImplementation(_entity, ComponentTypeT::typeId,
-      &_data);
+  auto updateData = this->CreateComponentImplementation(_entity,
+      ComponentTypeT::typeId, &_data);
+  auto comp = this->Component<ComponentTypeT>(_entity);
+  if (updateData)
+  {
+    if (!comp)
+    {
+      ignerr << "Internal error. Failure to create a component of type "
+        << ComponentTypeT::typeId << " for entity " << _entity
+        << ". This should never happen!\n";
+      return comp;
+    }
+    *comp = _data;
```

Comment:
> Isn't `CreateComponentImplementation` already setting the data?

It's not :frowning_face: since the data is part of the derived component class, it can't be set in `CreateComponentImplementation` because `CreateComponentImplementation` only has access to the `BaseComponent` class. So, we have to update the data in the template method that calls `CreateComponentImplementation`, since the template method has access to the derived class and its data. There's a comment that explains this: https://github.com/ignitionrobotics/ign-gazebo/blob/f0a0fe85029fe448b1c12ca522db8ce6f67dc71d/src/EntityComponentManager.cc#L671-L674